### PR TITLE
Create CSVReader from InputStream

### PR DESCRIPTION
**Pull request by josh**
Hey @dehesa, had another idea for a feature. 

Specifically, I was looking for some way to create a `CSVReader` reader from stdin.

Since reading from a URL was already implemented on top of `InputStream`, it seemed like exposing a public API to accept streams directly could work.

```swift
CSVReader(input: InputStream(fileAtPath: "/dev/stdin"))
```

I also saw there's a `FileHandle` API could maybe accomplish something similar. But I'm not sure how we'd implement converting a `FileHandle` -> `InputStream`

```swift
CSVReader(input: FileHandle.standardInput)
```

I'm just noting this as a possible alternative direction. I'm not too familiar with the differences between the APIs. Please share any historical knowledge if you have any 😄 

This PR is a draft attempt at the `InputStream` approach. Will add tests and clean it up if you think it's a good idea.

Thanks!
@josh

### CODE REVIEW

**Comment by dehesa** [comment]
Hi @josh,

Adding an API to read from an `InputStream` is definitely useful. It would enable the user to read from the console, network sockets, etc. Reading from the console's standard input is specially useful if you want to use `CodableCSV` in command-line applications.

As you have noticed, the `CSVReader` initializer which accepts a file URL does use `InputStream` internally. That stream is then forwarded to the decoder's `CSVReader._StreamBuffer`, where it is used.

When I was writing the API, I didn't expose the `InputStream` on purpose. The functionality to read from a file (which you know the end) is different from reading from a stream (where you receive many _pauses_/ends and maybe a final disconnection). Currently, `CSVReader._StreamBuffer` only supports the file-reading functionality with the added optimization to read in chunks (to not be calling down the kernel often).

If we want to add support for `stdin`, we would need to:
1. Pick the lower level APIs which suit our needs.
   
   Whether [`readLine(strippingNewline:)`](https://blog.faodailtechnology.com/reading-standard-input-stdin-in-swift-45f81b3caf2c), [`InputStream(fileAtPath: "/dev/stdin")`](https://stackoverflow.com/questions/29302904/get-input-from-stdin-using-nsinputstream-in-swift), or [`FileHandle.standardInput`](https://stackoverflow.com/questions/49748507/listening-to-stdin-in-swift). There might be more I am unaware of.

2. Create a new `makeDecoder` in the `ReaderDecoder.swift` file and/or stream buffer.

   The buffer will fetch whatever is available in `stdin` and maybe listen for more?

3. Verify how `CSVReader` works with this new _reading paradigm_.

   Specifically, what happens when a request for a new row occurs, but the user hasn't type the row yet (or the piped program hasn't send the answer yet). I guess the correct behavior would be to "freeze-and-wait". In any case, we should test all this and maybe throw new errors.

How does that sound?

**Comment by josh** [comment]
Ah, okay, so you're thinking reader initializer call that's specific to stdin. How do you think that could be named? Maybe it's like a special case stream enum? `CSVReader(input: .stdin)`

> Whether `readLine(strippingNewline:)`, `InputStream(fileAtPath: "/dev/stdin")`, or `FileHandle.standardInput`. There might be more I am unaware of.

I'll give `InputStream` an initial go since there's already support for that.

As for `readLine`:

> Standard input is interpreted as UTF-8. Invalid bytes are replaced by Unicode replacement characters.

It would lock you into UTF-8 which may not be a limitation we want.

> Specifically, what happens when a request for a new row occurs, but the user hasn't type the row yet (or the piped program hasn't send the answer yet). I guess the correct behavior would be to "freeze-and-wait". In any case, we should test all this and maybe throw new errors.

It looks like `InputStream` and `FileHandle` both have non-blocking options via blocks. I don't personally have a use-case for async reads, though that could be an API for later.

I think "freeze-and-wait" blocking aligns nicely with the iterative row reader. I'd like to ensure that either reading from a file or stdin acts the same.

```swift
// $ parse-csv path/to/file.csv
let reader = try CSVReader(input: URL(filename))

// $ cat path/to/file.csv | parse-csv
let reader = try CSVReader(input: .stdin)

// handle filename or stdin
for row in reader { print(row) }
```

File: sources/imperative/reader/ReaderAPI.swift [hunk 1]
```diff
@@ -71,6 +71,46 @@ extension CSVReader {
             try self.init(configuration: configuration, buffer: buffer, decoder: decoder)
         }
     }
+
+    public enum Stream {
+        case stdin
+    }
```
Comment (josh) [inline sources/imperative/reader/ReaderAPI.swift:77]: What file should this enum go in?

File: sources/imperative/reader/ReaderAPI.swift [hunk 2]
```diff
@@ -71,6 +71,46 @@ extension CSVReader {
...
+    /// - parameter input: The stream type to be parsed.
+    /// - parameter configuration: Recipe detailing how to parse the CSV data (i.e. encoding, delimiters, etc.).
+    /// - throws: `CSVError<CSVReader>` exclusively.
+    public convenience init(input: Stream, configuration: Configuration = .init()) throws {
+        if configuration.presample {
+            // A. If the `presample` configuration has been set, read all data from STDIN.
+            try self.init(input: FileHandle.standardInput.readDataToEndOfFile(), configuration: configuration); return
+        } else {
+            // B. Otherwise, create an input stream and start parsing byte-by-byte.
+            guard let stream = InputStream(fileAtPath: "/dev/stdin") else { throw Error._invalidStdin() }
```
Comment (josh) [inline sources/imperative/reader/ReaderAPI.swift:91]: Unsure if this should be a throwing error or a `fatalError` if stdin isn't available. I guess I'm not sure how it's reachable. I think even a closed stdin will still be able to construct a stream for. It will just be `streamStatus == .closed`.

File: sources/imperative/reader/ReaderAPI.swift [hunk 3]
```diff
@@ -71,6 +71,46 @@ extension CSVReader {
...
+            let (encoding, unusedBytes): (String.Encoding, [UInt8])
+            do {
+                // B.2. Check whether the input data has a BOM.
+                let inferred = try String.Encoding.infer(from: stream)
+                // B.3. Select the appropriate encoding depending from the user provided encoding (if any), and the BOM encoding (if any).
+                encoding = try CSVReader.selectEncodingFrom(provided: configuration.encoding, inferred: inferred.encoding)
+                unusedBytes = inferred.unusedBytes
+            } catch let error {
+                if stream.streamStatus != .closed { stream.close() }
+                throw error
```
Comment (josh) [inline sources/imperative/reader/ReaderAPI.swift:105]: Most of this body is similar to `init` with `Data`. Maybe extract a private method?

**Comment by josh** [comment]
Here's some scripts I'm testing with.

```swift
// usage: slow-csv <count>
import Darwin

let count = CommandLine.arguments.count < 2 ? 1 : Int(CommandLine.arguments[1])!

for n in 1 ... count {
    usleep(1_000_000)
    print("\(n),\(n + 1),\(n + 2)")
}
```

```swift
// usage: read-csv <count> <filename>

import CodableCSV
import Foundation

let start = DispatchTime.now()

let count: Int
let reader: CSVReader

if CommandLine.arguments.count < 2 {
    count = 1
    reader = try CSVReader(input: .stdin)
} else if CommandLine.arguments.count < 3 {
    count = Int(CommandLine.arguments[1])!
    reader = try CSVReader(input: .stdin)
} else {
    count = Int(CommandLine.arguments[1])!
    let url = URL(fileURLWithPath: CommandLine.arguments[2])
    reader = try CSVReader(input: url)
}

for row in reader.prefix(count) {
    let elapsed = Double(DispatchTime.now().uptimeNanoseconds - start.uptimeNanoseconds) / 1_000_000_000
    print("\(String(format: "%.3f", elapsed))s: \(row)")
}
```

Producer generates a row every second up to 10 total. The consumer reads the first 5 lines.

```
$ ./slow-csv.swift 10 | ./read-csv.swift 5
```

I think this is an interesting test case as it should be possible at best for the consumer to finish in 5 seconds (1 second per line) and also print 1 line every second. At the worst case with no streaming, all the lines will be printed to the screen at once after waiting 10 seconds.

**Comment by dehesa** [comment]
Wait, do your scripts work? 😱 I thought some deeper changes in the way streams are read would be required. That is great!

> Ah, okay, so you're thinking reader initializer call that's specific to `stdin`. How do you think that could be named? Maybe it's like a special case stream enum? `CSVReader(input: .stdin)`

What you propose is quite elegant and friendly for people that don't know much about streams. However if we can assure the correct work of any `InputStream` (whether `.stdin` or network sockets, or whatever), maybe it is worth to expose `InputStream`. I do quite like your syntax, though. Maybe overloading `InputStream` with a static property? I am not sure...

```swift
extension CSVReader {
    public convenience init(input: InputStream, configuration: Configuration) throws { ... }
}

extension InputStream {
    public static standardInput: InputStream { ... }
}
```

> It looks like InputStream and FileHandle both have non-blocking options via blocks. I don't personally have a use-case for async reads, though that could be a neat API for later.

I don't either and I believe the async reads would require deeper code changes. We can skip that for now.

> I think "freeze-and-wait" blocking aligns nicely with the iterative row reader.

I think so too, but since I don't have tons of experience with `InputStream`, I am a bit wary of what happens on edge cases (or not so edge cases), such as the row delimiter is not a line feed or the user is escaping a field and then press enter in the command-line. How does the `CSVReader` behaves?

Addressing your scripts, they seem to test the regular usage (which is a good first step). I haven't been able to run them though, since CodableCSV library is not imported. How do you run them? I add the shebang `#! /usr/bin/swift` at the top and even try to do the `-I moduleDirectory`, but I am getting fatal errors from the compiler. Could you provide some guidance?

File: sources/imperative/reader/ReaderAPI.swift (hunk 1 above)
Comment (dehesa) [reply sources/imperative/reader/ReaderAPI.swift:77]: If we decide to go in this direction, I would add it into `ReaderCollections.swift` and maybe rename the file to `ReaderIO.swift` or something like that.

File: sources/imperative/reader/ReaderAPI.swift (hunk 2 above)
Comment (dehesa) [reply sources/imperative/reader/ReaderAPI.swift:91]: I prefer to throw errors if the library user can recover from it and do something instead. For the specific situation of `stdin`: what is the `streamStatus` right after you create it? Do you have to open it?

If we decide to expose `InputStream` we could have a check and let pass anything that is not `.error`, or `.closed`, or `.atEnd`.

File: sources/imperative/reader/ReaderAPI.swift (hunk 3 above)
Comment (dehesa) [reply sources/imperative/reader/ReaderAPI.swift:105]: You  mean `init(input: URL)`? If we expose `InputStream`, we can make `init(input: URL)` a convenience for `init(input: InputStream)`.

File: sources/imperative/reader/ReaderAPI.swift [hunk 4]
```diff
@@ -220,6 +271,11 @@ fileprivate extension CSVReader.Error {
               help: "Make sure the URL is valid and you are allowed to access the file. Alternatively set the configuration's presample or load the file in a data blob and use the reader's data initializer.",
               userInfo: ["File URL": url])
     }
+    /// Error raised when an input stream cannot be created for "/dev/stdin".
+    static func _invalidStdin() -> CSVError<CSVReader> {
+        .init(.streamFailure,
+              reason: "Creating an input stream from stdin failed.")
```
Comment (dehesa) [inline sources/imperative/reader/ReaderAPI.swift:277]: Try to add `help` cues. Even if they are a bit stupid, such as "Please retry again".

File: sources/imperative/reader/ReaderAPI.swift (hunk 2 above)
Comment (josh) [reply sources/imperative/reader/ReaderAPI.swift:91]: Looks like it's starts not open, same as a regular file's input stream.

```swift
let stream = InputStream(fileAtPath: "/dev/stdin")!
stream.streamStatus // .notOpen
stream.open()
stream.streamStatus // .open
```

> If we decide to expose InputStream we could have a check and let pass anything that is not .error, or .closed, or .atEnd.

Oh, I thought you wanted to avoid that. Happy to have either.

File: sources/imperative/reader/ReaderAPI.swift (hunk 3 above)
Comment (josh) [reply sources/imperative/reader/ReaderAPI.swift:105]: Sorry, yeah, I think was confused on if you wanted to have `InputStream` public.

**Comment by josh** [comment]
> Wait, do your scripts work? 😱 I thought some deeper changes in the way streams are read would be required.

Oh yes, your underlying code works 😁 

> Addressing your scripts, they seem to test the regular usage (which is a good first step). I haven't been able to run them though, since CodableCSV library is not imported. How do you run them? I add the shebang `#! /usr/bin/swift` at the top and even try to do the `-I moduleDirectory`, but I am getting fatal errors from the compiler. Could you provide some guidance?

I use [swift-sh](https://github.com/mxcl/swift-sh) to make things either. Otherwise you can make a dummy `Package.swift` and point to your local project. This is basically what swift-sh does.

**Comment by josh** [comment]
> I think so too, but since I don't have tons of experience with `InputStream`, I am a bit wary of what happens on edge cases (or not so edge cases), such as the row delimiter is not a line feed or the user is escaping a field and then press enter in the command-line. How does the `CSVReader` behaves?

Let me test to be sure. But the row delimiter issue should be fine. `InputStream` isn't encoding aware so it's not treating line feeds any differently that they would be from reading from a file. If you typed invalid CSV directly in the terminal you should get the same thrown decoding error.

**Comment by josh** [comment]
> Using my test scripts, there’s a few inefficiencies mostly related to the large chunk size. Need to see if there’s some equivalent to `read` `O_NONBLOCK` so we could read data up to the chunk size and not block waiting for the entire chunk to be available.

Hmm, so `InputStream` nor `FileHandler` have any partial read APIs exposed. If we wanted to support that'd we'd need to drop down to `open`/`read`/`select`. Maybe out of scope for this PR? Though, one of your ideas was to "Make Foundation optional". We research nonblocking after that.

**Comment by josh** [comment]
> I think so too, but since I don't have tons of experience with `InputStream`, I am a bit wary of what happens on edge cases (or not so edge cases), such as the row delimiter is not a line feed or the user is escaping a field and then press enter in the command-line. How does the `CSVReader` behaves?

✅ Tested `$0.delimiters.row = "\t"`
✅ Tested `a,"bad escape,c` that should throw `_invalidEscapedField`

File: sources/imperative/ReaderUtils.swift [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        return InputStream(fileAtPath: "/dev/stdin")!
+    }
+}
+
+internal extension Data {
+    /// Initialize Data by reading entire input stream. May throw a stream error.
+    /// - parameter stream: Stream to read.
+    /// - parameter chunk: Chunk size.
+    /// - throws: An NSError object representing the stream error.
+    init(stream: InputStream, chunk: Int) throws {
```
Comment (josh) [inline sources/imperative/ReaderUtils.swift:15]: What file should this extension be in?

File: sources/imperative/ReaderUtils.swift [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import Foundation
+
+extension InputStream {
+    /// Stream for reading from stdin.
+    public static var stdin: InputStream {
```
Comment (josh) [inline sources/imperative/ReaderUtils.swift:5]: What file should this extension be in?

Comment (josh) [reply sources/imperative/ReaderUtils.swift:5]: TIL I learned you can't have the same file name in the same target. I originally tried naming this file `Utils.swift`.

```
$ swift build
<unknown>:0: error: filename "Utils.swift" used twice: 'sources/declarative/Utils.swift' and 'sources/i<unknown>:0: note: filenames are used to distinguish private declarations with the same name
```

File: sources/imperative/reader/ReaderAPI.swift [hunk 7]
```diff
@@ -49,6 +49,21 @@ extension CSVReader {
...
+        }
+    }
+
+    /// Creates a reader instance that will be used to parse the given CSV stream.
+    ///
+    /// If the configuration's encoding hasn't been set and the input data doesn't contain a Byte Order Marker (BOM), UTF8 is presumed.
+    /// - parameter input: The stream to be parsed.
+    /// - parameter configuration: Recipe detailing how to parse the CSV data (i.e. encoding, delimiters, etc.).
+    /// - throws: `CSVError<CSVReader>` exclusively.
+    public convenience init(input stream: InputStream, configuration: Configuration = .init()) throws {
```
Comment (josh) [inline sources/imperative/reader/ReaderAPI.swift:62]: Alright, reverted my change back to what I originally had. The public interface for `InputStream` is back.

File: sources/imperative/ReaderUtils.swift [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        return InputStream(fileAtPath: "/dev/stdin")!
```
Comment (josh) [inline sources/imperative/ReaderUtils.swift:6]: This force unwraps here. But maybe we should `fatalError` with a custom message?

I suppose it could fail on non-unix platforms or if your binary was in a chroot'd fs (like a docker container) and there's just no `/dev/stdin` mounted.

**Comment by dehesa** [comment]
Ok, I got your scripts running and they do indeed work! 😃 After some head scratching asking myself how comes everything works that well, I reached the conclusion that everything is very synchronous (as you kindly suggested); therefore it is basically the same functionality as with `URL`. For example: `slow-csv 100 | read-csv 2` will first finish the 100 iterations and then just read two rows (I guess it is obvious with hindsight).

For this PR we can ignore any type of asynchronous operation, since most people won't use it that way. Although it would be nice to research that in the future (I will add it to the [Ideas project](https://github.com/dehesa/CodableCSV/projects/3)) and we should also document that behavior.

> Maybe out of scope for this PR? Though, one of your ideas was to "Make Foundation optional". We research nonblocking after that.

I would like to make the Foundation dependency optional. Foundation is not very popular in the _server-side_ and many projects directly avoid it. Also now that Swift is getting some traction on the Machine Learning community (i.e. Linux and Windows), it would be desirable to make CodableCSV the go-to library for reading/writing CSVs in ML pipelines. For that we need to remove the Foundation requirement and improve performance. You can see in [the wiki](https://github.com/dehesa/CodableCSV/wiki/Implicit-dependencies) all the current Foundation dependencies. But I digress; we will think about it once we cross that bridge.

Regarding the code placement and the `fatalError` message. I am happy to accept the PR as it is right now and I will figure it out (probably tomorrow) once I do the [Encoder/Decoder auditing](https://github.com/dehesa/CodableCSV/projects/1).

Do you want to add something else or should I merge it as it is?

**Comment by josh** [comment]
> Do you want to add something else or should I merge it as it is?

Sure! I'm down to keep iterating on this idea. I'll let you know if I come up any ideas to improve the processing efficiency.

> I would like to make the Foundation dependency optional.

Neat! I'd love to help with this as well.

**Comment by dehesa** [comment]
Thank you for your code and readiness @josh. I appreciate it.

**Comment by dehesa** [comment]
Hey @josh, just as FYI, I have performed the following additions/changes on top of your code:
1. `CSVReader`: add a static `decode(input:configuration:)` supporting `InputStream`.
2. `CSVDecoder`: add support for `InputStream`.
3. `InputStream`: change the `stdin` property name to `standardInput`.
4. `CSVReader`: wrap the `Data(stream:chunk:)` error into a `CSVError`.

I am not 100% convinced on point 3. Swift developers recommend to use full names whenever possible and Apple uses the name `standardInput` in their `FileHandle` API. So I went with that, but the short `stdin` has a place in my heart 😄 Do you have any strong opinions?

**Comment by josh** [comment]
Nice changes! I could go either way on stdin and standardInput. Both seem great for their own reasons.

**Comment by josh** [comment]
btw, published little CLI tool I just built on this API. https://github.com/josh/csv2json

🙌🏻

**Comment by dehesa** [comment]
It looks pretty cool. Congrats! and you are also using the new Argument Parser. That is in my things-to-learn box.

Let me know if I can help you out with anything.
